plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,633 @@
+
+static void
+fu_algoltek_usb_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FU_DEVICE_CLASS(fu_algoltek_usb_device_parent_class)->to_string(device, idt, str);
+}
+
+static guint8
+checkSum(GByteArray *byteArray)
+{
+	guint8 result = 0;
+	for (guint32 i = 0; i < byteArray->len; i++) {
+		result += (guint8)(byteArray->data[i] & 0xff);
+	}
+	result = (guint8)(~result + 1);
+	return result;
+}
+
+static GByteArray *
+fu_algoltek_device_CMD02(FuAlgoltekUsbDevice *self, int address, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+
+	/* first byte is length */
+	fu_byte_array_set_size(buf, 11, 0x0);
+	buf->data[0] = 5;
+	buf->data[1] = ALGOLTEK_USB_REQUEST_CMD02;
+	buf->data[2] = ((address >> 8) & 0xFF);
+	buf->data[3] = (address & 0xFF);
+	buf->data[10] = checkSum(buf);
+	if (!g_usb_device_control_transfer(usb_device,
+					   G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_INTERFACE,
+					   ALGOLTEK_USB_REQUEST_CMD02,
+					   address, /* value */
+					   0xFFFF,  /* idx */
+					   buf->data,
+					   buf->len,
+					   NULL,
+					   ALGOLTEK_DEVICE_USB_TIMEOUT,
+					   NULL,
+					   error)) {
+		g_prefix_error(error, "failed to contact device: ");
+		return NULL;
+	}
+
+	/* success */
+	return g_steal_pointer(&buf);
+}
+
+static GByteArray *
+fu_algoltek_device_CMD01(FuAlgoltekUsbDevice *self, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+	g_autoptr(GByteArray) versionData = g_byte_array_new();
+	gsize actual_length = 0;
+
+	/* first byte is length */
+	fu_byte_array_set_size(buf, 64, 0x0);
+	buf->data[0] = 3;
+	buf->data[1] = ALGOLTEK_USB_REQUEST_CMD01;
+	buf->data[63] = checkSum(buf);
+	if (!g_usb_device_control_transfer(usb_device,
+					   G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_INTERFACE,
+					   ALGOLTEK_USB_REQUEST_CMD01,
+					   0xFFFF, /* value */
+					   0xFFFF, /* idx */
+					   buf->data,
+					   buf->len,
+					   &actual_length,
+					   ALGOLTEK_DEVICE_USB_TIMEOUT,
+					   NULL,
+					   error)) {
+		g_prefix_error(error, "failed to contact device: ");
+		return NULL;
+	}
+
+	if (actual_length != buf->len) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "got 0x%x but requested 0x%x",
+			    (guint)actual_length,
+			    (guint)buf->len);
+		return NULL;
+	}
+
+	/* Remove cmd and length */
+	for (guint32 i = 2; i < buf->len; i++) {
+		if (buf->data[i] < 128)
+			fu_byte_array_append_uint8(versionData, buf->data[i]);
+	}
+	/* success */
+	return g_steal_pointer(&versionData);
+}
+
+static guint
+fu_algoltek_readout_info_int(GByteArray *backData)
+{
+	int backDataAddress = (backData->data[2] << 8) | backData->data[3];
```

Comment:
This isn't endian safe -- I think you either need to use `fu_struct_whatever_parse()` here or `fu_memread_uint16`

Reply:
Updated! I used `fu_memread_uint16( )`.